File: python-midonetclient/src/midonetclient/resource_base.py
# Copyright 2013, 2014 Midokura SARL, All Rights Reserved.
# All Rights Reserved
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.


class ResourceBase(object):

    media_type = 'application/json'  # default media_type for all subclasses

    def __init__(self, uri, dto, auth):
        self.uri = uri
        self.dto = dto
        self.auth = auth

    def _ensure_content_type(self, headers):
        """Ensure that http header dict has a value for 'Content-Type'"""
        headers.setdefault('Content-Type', self.media_type)

    def _ensure_accept(self, headers):
        """Ensure that http header dict has a value for 'Accept'"""
        headers.setdefault('Accept', self.media_type)

    def create(self, headers=None):
        """Does REST POST at some uri followed by REST GET at new location"""

        headers = headers or dict()
        self._ensure_content_type(headers)

        resp, _ = self.auth.do_request(self.uri, 'POST', body=self.dto,
                                       headers=headers)

        self._ensure_accept(headers)
        _, self.dto = self.auth.do_request(resp['location'], 'GET',
                                           headers=headers)
        return self

    def upload(self, uri, body, headers=None):
        """Does REST POST with a binary body at some URI"""

        headers = headers or dict()
        self._ensure_content_type(headers)

        resp, self.dto = self.auth.do_upload(uri, body=body, headers=headers)
        self._ensure_accept(headers)

        return self

    def get(self, headers=None, **kwargs):
        """Does REST GET at some uri"""
        headers = headers or dict()
        self._ensure_accept(headers)
        uri = self.dto['uri']
        _, self.dto = self.auth.do_request(uri, 'GET', headers=headers)
        return self

    def get_children(self, uri, query, headers, clazz, extra_args=None):
        """Does GET at uri and create objects found in server answer"""
        self._ensure_accept(headers)

        _, dtos = self.auth.do_request(uri, 'GET', query=query,
                                       headers=headers)

        return map(
            lambda dto: clazz(uri, dto, self.auth, *(extra_args or [])),
            dtos or []  # in case API returns empty when no hosts
        )

    def get_uri(self):
        """Return one's own uri"""
        return self.dto['uri']

    def update(self, headers=None):
        """Does PUT at own uri followed by GET at own uri"""
        headers = headers or dict()

        self._ensure_content_type(headers)

        self.auth.do_request(self.dto['uri'], 'PUT',
                             body=self.dto,
                             headers=headers)

        headers['Accept'] = self.media_type
        _, self.dto = self.auth.do_request(self.dto['uri'], 'GET',
                                           headers=headers)
        return self

    def delete(self, headers={}):
        """Does DELETE at own uri"""
        self.auth.do_request(self.dto['uri'], 'DELETE', headers=headers)
        return None

    def __repr__(self):
        return self.__class__.__name__ + str(self.dto)

File: cluster/midonet-cluster/src/main/java/org/midonet/client/resource/ResourceBase.java
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.client.resource;

import java.lang.reflect.Array;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.net.URI;
import java.util.ArrayList;
import java.util.Objects;

import javax.ws.rs.core.MultivaluedMap;
import javax.ws.rs.core.UriBuilder;

import org.midonet.client.WebResource;

/**
 * Base class for resource classes
 *
 * @param <T> concrete class
 * @param <U> type of principal DTO
 */
public abstract class ResourceBase<T extends ResourceBase<T,U>, U> {

    final WebResource resource;
    U principalDto;
    final String mediaType;
    final URI uriForCreation;

    public ResourceBase(WebResource resource, URI uriForCreation,
                        U principalDto, String mediaType) {
        this.resource = resource;
        this.uriForCreation = uriForCreation;
        this.principalDto = principalDto;
        this.mediaType = mediaType;

    }

    /**
     * Gets URI of the resource
     *
     * @return URI of the resource
     */
    public abstract URI getUri();

    /**
     * Returns DTO of the resource. This doesn't invold REST API call, instead
     * returns principal DTO object so the client can get raw data.
     *
     * @return DTO of the resource
     */
    public U getLocalDto() {
        return principalDto;
    }

    <T, U> ResourceCollection<T>
    getChildResources(URI uri,
                      MultivaluedMap<String, String> queryParams,
                      String collectionMediaType,
                      Class<T> resourceClazz,
                      Class<U> newPrincipalDtoClazz) {
        return getChildResources(uri, queryParams, collectionMediaType, null,
                                 resourceClazz, newPrincipalDtoClazz);
    }

    /**
     * Get resources that belong to the resource.
     *
     * @param uri                  URI for getting resource(s)
     * @param queryParams          Query parameters; null for no queries
     * @param collectionMediaType  media type for collection of the resource
     * @param elementMediaType     media type for the element itself, if
     *                             resourceClazz' constructor needs it.
     * @param resourceClazz
     * @param newPrincipalDtoClazz
     * @param <T>                  Type of newly returned resources
     * @param <U>                  Type of DTO of newly returned resources
     * @return
     */
    <T, U> ResourceCollection<T>
    getChildResources(URI uri,
                      MultivaluedMap<String, String> queryParams,
                      String collectionMediaType,
                      String elementMediaType,
                      Class<T> resourceClazz,
                      Class<U> newPrincipalDtoClazz) {
        ResourceCollection<T> container =
            new ResourceCollection<T>(new ArrayList<T>());

        Class<?> newPrincipalDtoArrayClazz =
            Array.newInstance(newPrincipalDtoClazz, 0).getClass();

        // newPrincipalDtoArrray should be T[]
        @SuppressWarnings("unchecked")
        T[] dtoArray = resource.get(uri,
                                    queryParams,
                                    (Class<T[]>) newPrincipalDtoArrayClazz,
                                    collectionMediaType);


        try {
            if (elementMediaType == null) {
                Constructor<T> ctor = resourceClazz.getConstructor(
                    WebResource.class,
                    URI.class,
                    newPrincipalDtoClazz);
                for (T e : dtoArray) {
                    container.add(ctor.newInstance(resource, uri, e));
                }
            } else {
                Constructor<T> ctor = resourceClazz.getConstructor(
                        WebResource.class,
                        URI.class,
                        newPrincipalDtoClazz,
                        String.class);
                for (T e : dtoArray) {
                    container.add(ctor.newInstance(resource, uri, e,
                                                   elementMediaType));
                }
            }
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        } catch (InstantiationException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }
        return container;
    }

    /**
     * Given the uri, returns updated this resource after getting resource data
     * from API server.
     *
     * @param uri URI of the resource
     * @return this
     */
    public T get(URI uri) {
        if (uri == null)
            throw new IllegalArgumentException("The URI can't be null.");

        // principalDto should be type U as ctor defines.
        @SuppressWarnings("unchecked")
        Class<U> clazz = (Class<U>) principalDto.getClass();

        principalDto = resource.get(uri, null, clazz, mediaType);

        // "this" should be type T which is subtype of this class.
        @SuppressWarnings("unchecked")
        T updated = (T) this;
        return updated;
    }

    /**
     * Returns this after getting resource data from API server
     *
     * @return this
     */
    public T get() {
        return get(getUri());
    }

    /**
     * Creates a resource on API server by POSTing to the URI that is stored in
     * uriForCreation on the API server.
     *
     * @return this
     */
    public T create() {
        URI location = resource.post(uriForCreation, principalDto,
                                     mediaType);
        return get(location);
    }

    /**
     * Update the resource on API server with local DTO
     *
     * @return this
     */
    public T update() {
        resource.put(getUri(), principalDto, mediaType);
        return get(getUri());
    }

    /**
     * Delete the resource on API server
     */
    public void delete() {
        resource.delete(getUri());
    }

    /**
     * Create a URI object from a URI template, token string and the replacement
     * value.
     *
     * @param template URI template string
     * @param token Token to replace
     * @param value Value to replace the token with
     * @return URI with token replaced with value
     */
    protected URI createUriFromTemplate(String template, String token,
                                        Object value) {
        return UriBuilder.fromUri(
                template.replace(token, value.toString())).build();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this)
            return true;

        if (obj == null || getClass() != obj.getClass())
            return false;

        @SuppressWarnings("unchecked") // safe at this point
        ResourceBase<T, U> otherResource = (ResourceBase<T, U>)obj;

        return Objects.equals(principalDto, otherResource.principalDto) &&
               Objects.equals(mediaType, otherResource.mediaType);
    }
}
